README.md
[![Github top language](https://img.shields.io/github/languages/top/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](../../)
[![Codacy grade](https://img.shields.io/codacy/grade/[proj-id].svg?style=for-the-badge)](https://www.codacy.com/gh/FHPythonUtils/MiniEncoding)
[![Repository size](https://img.shields.io/github/repo-size/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](../../)
[![Issues](https://img.shields.io/github/issues/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](../../issues)
[![License](https://img.shields.io/github/license/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](/LICENSE.md)
[![Commit activity](https://img.shields.io/github/commit-activity/m/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](../../commits/master)
[![Last commit](https://img.shields.io/github/last-commit/FHPythonUtils/MiniEncoding.svg?style=for-the-badge)](../../commits/master)
[![PyPI Downloads](https://img.shields.io/pypi/dm/miniencoding.svg?style=for-the-badge)](https://pypi.org/project/miniencoding/)
[![PyPI Version](https://img.shields.io/pypi/v/miniencoding.svg?style=for-the-badge)](https://pypi.org/project/miniencoding/)

<!-- omit in TOC -->
# MiniEncoding

<img src="readme-assets/icons/name.png" alt="Project Icon" width="750">


Ascii representations. Stores ascii as a bitarray in ascii8 form.


## Encodings

### miniencoding.asciibytes AsciiBytes

| Encoding    |  To   | From  |
| :---------- | :---: | :---: |
| Ascii8      |   ✔   |   ✔   |
| Ascii7      |   ✔   |   ✔   |
| DecSixBit   |   ✔   |   ✔   |
| ICL*        |   ✔   |   ✔   |
| SixBitAscii |   ✔   |   ✔   |


```txt
* ↑ and	← not supported instead uses ^ and _
```

### miniencoding.tables

Encode to a charset with:
```python
def toCharset(charset, string, toUpper=False):
	"""Convert unicode text to various charsets using a lookup table

	Args:
		charset (string): lookup table
		string (string): unicode string to convert
		toUpper (bool, optional): Make chars uppercase before converting
		(intended for 6 bit charsets or charsets that do not support uppercase).
		Defaults to False.

	Returns:
		bytes: sequence of bytes (split into bits of length
	"""
```

Decode from a charset to Unicode with:
```python
def toUnicode(charset, bytestream):
	"""Convert text encoded with various charsets to unicode using a charset
	lookup table

	Args:
		charset (string): lookup table
		bytestream (bytes): sequence of bytes (split into bits of length
		log2(len(charset))) to convert

	Returns:
		string: unicode string
	"""
```

Available charsets (or make your own):
```python
# CDC 1604: Magnetic tape BCD codes
CDC1604_MAGTAPE
# CDC 1604: Punched card codes
CDC1604_PUNCHCARD
# CDC 1612: Printer codes (business applications)
CDC1612
# DEC SIXBIT
DEC_SIXBIT
# ECMA-1
EMCA1
# ICL Mainframes
ICL
# SixBit ASCII (used by AIS)
SIXBIT
# GOST 10859 § 6-bit code: with only Cyrillic upper case letters
GOST
# GSM 7-bit default alphabet and extension table of 3GPP TS 23.038 / GSM 03.38
GSM7
# ASCII
ASCII7
# IBM 48-character BCDIC code
IBM48
# IBM 704 BCD code
IBM704
# IBM 7090/7094 character set
IBM7090
# IBM 1401 BCD code
IBM1401
# GBCD code
GBCD
# Burroughs B5500 BCD code
BURROUGHS_B5500
# Code page 353
CP353
# Code page 355
CP355
# Code page 357
CP357
# Code page 358
CP358
# Code page 359/360
CP359
```

For more info see
https://en.wikipedia.org/wiki/Six-bit_character_code#Types_of_six-bit_codes

- [Encodings](#encodings)
	- [miniencoding.asciibytes AsciiBytes](#miniencodingasciibytes-asciibytes)
	- [miniencoding.tables](#miniencodingtables)
- [Install With PIP](#install-with-pip)
- [Language information](#language-information)
	- [Built for](#built-for)
- [Install Python on Windows](#install-python-on-windows)
	- [Chocolatey](#chocolatey)
	- [Download](#download)
- [Install Python on Linux](#install-python-on-linux)
	- [Apt](#apt)
- [How to run](#how-to-run)
	- [With VSCode](#with-vscode)
	- [From the Terminal](#from-the-terminal)
- [How to update, build and publish](#how-to-update-build-and-publish)
- [Download](#download-1)
	- [Clone](#clone)
		- [Using The Command Line](#using-the-command-line)
		- [Using GitHub Desktop](#using-github-desktop)
	- [Download Zip File](#download-zip-file)
- [Community Files](#community-files)
	- [Licence](#licence)
	- [Changelog](#changelog)
	- [Code of Conduct](#code-of-conduct)
	- [Contributing](#contributing)
	- [Security](#security)
	- [Support](#support)
	- [Rationale](#rationale)

## Install With PIP

```python
pip install miniencoding
```

Head to https://pypi.org/project/miniencoding/ for more info


## Language information
### Built for
This program has been written for Python 3 and has been tested with
Python version 3.8.0 <https://www.python.org/downloads/release/python-380/>.

## Install Python on Windows
### Chocolatey
```powershell
choco install python
```
### Download
To install Python, go to <https://www.python.org/> and download the latest
version.

## Install Python on Linux
### Apt
```bash
sudo apt install python3.8
```

## How to run
### With VSCode
1. Open the .py file in vscode
2. Ensure a python 3.8 interpreter is selected (Ctrl+Shift+P > Python:Select
Interpreter > Python 3.8)
3. Run by pressing Ctrl+F5 (if you are prompted to install any modules, accept)
### From the Terminal
```bash
./[file].py
```

## How to update, build and publish

1. Ensure you have installed the following dependencies
	Linux
	```bash
	wget dephell.org/install | python3.8
	wget https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python3.8
	```
	Windows
	```powershell
	(wget dephell.org/install -UseBasicParsing).Content | python
	(wget https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py -UseBasicParsing).Content | python
	```
2. Use poetry for the heavy lifting and dephell to generate requirements
	```bash
	poetry update
	dephell deps convert
	```
3. Build/ Publish
	```bash
	poetry build
	poetry publish
	```
	or
	```bash
	poetry publish --build
	```


## Download
### Clone
#### Using The Command Line
1. Press the Clone or download button in the top right
2. Copy the URL (link)
3. Open the command line and change directory to where you wish to
clone to
4. Type 'git clone' followed by URL in step 2
```bash
$ git clone https://github.com/FHPythonUtils/MiniEncoding
```

More information can be found at
<https://help.github.com/en/articles/cloning-a-repository>

#### Using GitHub Desktop
1. Press the Clone or download button in the top right
2. Click open in desktop
3. Choose the path for where you want and click Clone

More information can be found at
<https://help.github.com/en/desktop/contributing-to-projects/cloning-a-repository-from-github-to-github-desktop>

### Download Zip File

1. Download this GitHub repository
2. Extract the zip archive
3. Copy/ move to the desired location

## Community Files
### Licence
MIT License
Copyright (c) FredHappyface
(See the [LICENSE](/LICENSE.md) for more information.)

### Changelog
See the [Changelog](/CHANGELOG.md) for more information.

### Code of Conduct
Online communities include people from many backgrounds. The *Project*
contributors are committed to providing a friendly, safe and welcoming
environment for all. Please see the
[Code of Conduct](https://github.com/FHPythonUtils/.github/blob/master/CODE_OF_CONDUCT.md)
 for more information.

### Contributing
Contributions are welcome, please see the
[Contributing Guidelines](https://github.com/FHPythonUtils/.github/blob/master/CONTRIBUTING.md)
for more information.

### Security
Thank you for improving the security of the project, please see the
[Security Policy](https://github.com/FHPythonUtils/.github/blob/master/SECURITY.md)
for more information.

### Support
Thank you for using this project, I hope it is of use to you. Please be aware that
those involved with the project often do so for fun along with other commitments
(such as work, family, etc). Please see the
[Support Policy](https://github.com/FHPythonUtils/.github/blob/master/SUPPORT.md)
for more information.

### Rationale
The rationale acts as a guide to various processes regarding projects such as
the versioning scheme and the programming styles used. Please see the
[Rationale](https://github.com/FHPythonUtils/.github/blob/master/RATIONALE.md)
for more information.


CHANGELOG.md
# Changelog
All major and minor version changes will be documented in this file. Details of
patch-level version changes can be found in [commit messages](../../commits/master).

## 2020 - 2020/08/21
- First release
- Features ascii8, ascii7 ascii6 (DEC Six Bit, ICL, Six Bit ascii)
- Supports 10 6/7 bit encodings with miniencoding.tables

## 2020.1 - 2020/09/10
- Supports 21 6/7 bit encodings with miniencoding.tables (+11). Those are:
```python
# IBM 48-character BCDIC code
IBM48 = " 1234567890#@????/STUVWXYZ?,%???-JKLMNOPQR?$*???&ABCDEFGHI?.⌑???"
# IBM 704 BCD code
IBM704 = "0123456789?#@???&ABCDEFGHI?.⌑???-JKLMNOPQR?$*??? /STUVWXYZ?,%???"
# IBM 7090/7094 character set
IBM7094 = IBM7090 = "0123456789?=\"???&ABCDEFGHI0.)???-JKLMNOPQR0$*??? /STUVWXYZ±,(???"
# IBM 1401 BCD code
IBM1401 = " 1234567890#@:>√¢/STUVWXYZ‡,%='\"-JKLMNOPQR!$*);Δ&ABCDEFGHI?.⌑(<⯒"
# GBCD code
GBCD = "0123456789[#@:>? ABCDEFGHI&.](<\\^JKLMNOPQR-$*);'+/STUVWXYZ_,%=\"!"
# Burroughs B5500 BCD code
BURROUGHS_B5500 = "0123456789#@?:>≥+ABCDEFGHI.[&(<←×JKLMNOPQR$*-);≤ /STUVWXYZ,%≠=]\""
# Code page 353
CP353 = " 1234567890#@:>√␢/STUVWXYZ‡,%γ\\⧻-JKLMNOPQR!#*];Δ&ABCDEFGHI?.⌑[<⯒"
# Code page 355
CP355 = " 1234567890#????@/STUVWXYZ‡,?γ??-JKLMNOPQR<$????&ABCDEFGHI).????"
# Code page 357
CP357 = " 1234567890=????'/STUVWXYZ‡,????-JKLMNOPQR!$????+ABCDEFGHI?.????"
# Code page 358
CP358 = " 1234567890'????!/STUVWXYZ‡,????-JKLMNOPQR<;????=ABCDEFGHI>.????"
# Code page 359/360
CP360 = CP359 = " 1234567890#????@/STUVWXYZ?,????-JKLMNOPQR?$????&ABCDEFGHI?.????"
```


test_tables.py
"""Test the miniencoding lib

A decent testing approach is to test the round trip with a random, valid
string of bytes. by taking this approach, the same error/ bug would have to be
present in both the 'from' and 'to' functions which whilst possible is unlikely
"""
# pylint: disable=invalid-name
import random
import string
from miniencoding.tables import *


def test_CDC1604_MAGTAPE_len():
	""" Test CDC1604_MAGTAPE length """
	assert len(CDC1604_MAGTAPE) == 64


def test_CDC1604_MAGTAPE():
	""" Test CDC1604_MAGTAPE round trip """
	testString = "?1234567890#@??? /STUVWXYZ?,%???-JKLMNOPQR0$*???&ABCDEFGHI0.¤???"
	assert toUnicode(CDC1604_MAGTAPE, toCharset(CDC1604_MAGTAPE,
	testString)) == testString


def test_CDC1604_PUNCHCARD_len():
	""" Test CDC1604_PUNCHCARD length """
	assert len(CDC1604_PUNCHCARD) == 64


def test_CDC1604_PUNCHCARD():
	""" Test CDC1604_PUNCHCARD round trip """
	testString = "?1234567890=-??? /STUVWXYZ?,(???—JKLMNOPQR0$*???+ABCDEFGHI0.)???"
	assert toUnicode(CDC1604_PUNCHCARD, toCharset(CDC1604_PUNCHCARD,
	testString)) == testString


def test_CDC1612_len():
	""" Test CDC1612 length """
	assert len(CDC1612) == 64


def test_CDC1612():
	""" Test CDC1612 round trip """
	testString = ":1234567890=≠≤![ /STUVWXYZ],(→≡~—JKLMNOPQR%$*↑↓>+ABCDEFGHI<.)≥?;"
	assert toUnicode(CDC1612, toCharset(CDC1612, testString)) == testString


def test_DEC_SIXBIT_len():
	""" Test DEC_SIXBIT length """
	assert len(DEC_SIXBIT) == 64


def test_DEC_SIXBIT():
	""" Test DEC_SIXBIT round trip """
	testString = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_"
	assert toUnicode(DEC_SIXBIT, toCharset(DEC_SIXBIT, testString)) == testString


def test_EMCA1_len():
	""" Test EMCA1 length """
	assert len(EMCA1) == 64


def test_EMCA1():
	""" Test EMCA1 round trip """
	testString = " \t\n\v\f\r\x0e\x0f()*+,-./0123456789:;<=>?\x00ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]\x1b\x7f"
	assert toUnicode(EMCA1, toCharset(EMCA1, testString)) == testString


def test_ICL_len():
	""" Test ICL length """
	assert len(ICL) == 64


def test_ICL():
	""" Test ICL round trip """
	testString = "0123456789:;<=>? !\"#£%&'()*+,-./@ABCDEFGHIJKLMNOPQRSTUVWXYZ[$]↑←"
	assert toUnicode(ICL, toCharset(ICL, testString)) == testString


def test_SIXBIT_len():
	""" Test SIXBIT length """
	assert len(SIXBIT) == 64


def test_SIXBIT():
	""" Test SIXBIT round trip """
	testString = "@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_ !\"#$%&'()*+,-./0123456789:;<=>?"
	assert toUnicode(SIXBIT, toCharset(SIXBIT, testString)) == testString


def test_GOST_len():
	""" Test GOST length """
	assert len(GOST) == 64


def test_GOST():
	""" Test GOST round trip """
	testString = "0123456789+-/,. ⏨↑()×=;[]*‘’≠<>:АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЫЬЭЮЯ\x7f"
	assert toUnicode(GOST, toCharset(GOST, testString)) == testString


def test_GSM7_len():
	""" Test GSM7 length """
	assert len(GSM7) == 128


def test_GSM7():
	""" Test GSM7 round trip """
	testString = "@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ\x07ÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ§¿abcdefghijklmnopqrstuvwxyzäöñüà"
	assert toUnicode(GSM7, toCharset(GSM7, testString)) == testString


def test_ASCII7_len():
	""" Test ASCII7 length """
	assert len(ASCII7) == 128


def test_ASCII7():
	""" Test ASCII7 round trip """
	testString = bytes(range(0, 128)).decode("utf-8")
	assert toUnicode(ASCII7, toCharset(ASCII7, testString)) == testString


def test_IBM48_len():
	""" Test IBM48 length """
	assert len(IBM48) == 64


def test_IBM48():
	""" Test IBM48 round trip """
	testString = " 1234567890#@????/STUVWXYZ?,%???-JKLMNOPQR?$*???&ABCDEFGHI?.⌑???"
	assert toUnicode(IBM48, toCharset(IBM48, testString)) == testString


def test_IBM704_len():
	""" Test IBM704 length """
	assert len(IBM704) == 64


def test_IBM704():
	""" Test IBM704 round trip """
	testString = "0123456789?#@???&ABCDEFGHI?.⌑???-JKLMNOPQR?$*??? /STUVWXYZ?,%???"
	assert toUnicode(IBM704, toCharset(IBM704, testString)) == testString


def test_IBM7090_len():
	""" Test IBM7090 length """
	assert len(IBM7090) == 64


def test_IBM7090():
	""" Test IBM7090 round trip """
	testString = "0123456789?=\"???&ABCDEFGHI0.)???-JKLMNOPQR0$*??? /STUVWXYZ±,(???"
	assert toUnicode(IBM7090, toCharset(IBM7090, testString)) == testString


def test_IBM1401_len():
	""" Test IBM1401 length """
	assert len(IBM1401) == 64


def test_IBM1401():
	""" Test IBM1401 round trip """
	testString = " 1234567890#@:>√¢/STUVWXYZ‡,%='\"-JKLMNOPQR!$*);Δ&ABCDEFGHI?.⌑(<⯒"
	assert toUnicode(IBM1401, toCharset(IBM1401, testString)) == testString


def test_GBCD_len():
	""" Test GBCD length """
	assert len(GBCD) == 64


def test_GBCD():
	""" Test GBCD round trip """
	testString = "0123456789[#@:>? ABCDEFGHI&.](<\\^JKLMNOPQR-$*);'+/STUVWXYZ_,%=\"!"
	assert toUnicode(GBCD, toCharset(GBCD, testString)) == testString


def test_BURROUGHS_B5500_len():
	""" Test BURROUGHS_B5500 length """
	assert len(BURROUGHS_B5500) == 64


def test_BURROUGHS_B5500():
	""" Test BURROUGHS_B5500 round trip """
	testString = "0123456789#@?:>≥+ABCDEFGHI.[&(<←×JKLMNOPQR$*-);≤ /STUVWXYZ,%≠=]\""
	assert toUnicode(BURROUGHS_B5500, toCharset(BURROUGHS_B5500, testString)) == testString


def test_CP353_len():
	""" Test CP353 length """
	assert len(CP353) == 64


def test_CP353():
	""" Test CP353 round trip """
	testString = " 1234567890#@:>√␢/STUVWXYZ‡,%γ\\⧻-JKLMNOPQR!#*];Δ&ABCDEFGHI?.⌑[<⯒"
	assert toUnicode(CP353, toCharset(CP353, testString)) == testString


def test_CP355_len():
	""" Test CP355 length """
	assert len(CP355) == 64


def test_CP355():
	""" Test CP355 round trip """
	testString = " 1234567890#????@/STUVWXYZ‡,?γ??-JKLMNOPQR<$????&ABCDEFGHI).????"
	assert toUnicode(CP355, toCharset(CP355, testString)) == testString


def test_CP357_len():
	""" Test CP357 length """
	assert len(CP357) == 64


def test_CP357():
	""" Test CP357 round trip """
	testString = " 1234567890=????'/STUVWXYZ‡,????-JKLMNOPQR!$????+ABCDEFGHI?.????"
	assert toUnicode(CP357, toCharset(CP357, testString)) == testString


def test_CP358_len():
	""" Test CP358 length """
	assert len(CP358) == 64


def test_CP358():
	""" Test CP358 round trip """
	testString = " 1234567890'????!/STUVWXYZ‡,????-JKLMNOPQR<;????=ABCDEFGHI>.????"
	assert toUnicode(CP358, toCharset(CP358, testString)) == testString


def test_CP359_len():
	""" Test CP359 length """
	assert len(CP359) == 64


def test_CP359():
	""" Test CP359 round trip """
	testString = " 1234567890#????@/STUVWXYZ?,????-JKLMNOPQR?$????&ABCDEFGHI?.????"
	assert toUnicode(CP359, toCharset(CP359, testString)) == testString


test_asciibytes.py
"""Test the miniencoding lib

A decent testing approach is to test the round trip with a random, valid
string of bytes. by taking this approach, the same error/ bug would have to be
present in both the 'from' and 'to' functions which whilst possible is unlikely
"""
# pylint: disable=invalid-name
import random
import string
from miniencoding.asciibytes import AsciiBytes


def generator(length):
	"""Generate a random sequence of bytes

	Args:
		length (int): length

	Returns:
		bytes: random sequence of bytes
	"""
	letters = string.printable
	return ''.join(random.choice(letters) for i in range(length)).encode()


def test_ascii8():
	"""Ascii8 has no strange behaviours or edge cases so this should be pretty
	straightforward
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert testAscii.toAscii8() == testString


def test_ascii8Len():
	"""Ascii8 has a length of 1 byte per char
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert len(testAscii.toAscii8()) == 32


def test_ascii7Printable():
	"""Ascii7 has no strange behaviours or edge cases for printable chars so
	this should be pretty straightforward
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii2 = AsciiBytes()
	testAscii.fromAscii8(testString)
	testAscii2.fromAscii7(testAscii.toAscii7())
	assert testAscii2.toAscii8() == testString


def test_ascii7NonPrintable():
	"""ascii7 cannot process ascii values over 127, will replace chars with '?'
	"""
	testAscii = AsciiBytes()
	testAscii2 = AsciiBytes()
	testAscii.fromAscii8(bytes([127, 128]))
	testAscii2.fromAscii7(testAscii.toAscii7())
	assert testAscii2.toAscii8() == bytes([127]) + b"?"


def test_ascii7Len():
	"""Ascii7 has a length of 7/8 byte per char
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert len(testAscii.toAscii7()) == 4 * 7


def test_decSixBit():
	"""Dec Six Bit has edge cases and strange behaviours as shown by the
	inline for loop for the variable 'expected'
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii2 = AsciiBytes()
	testAscii.fromAscii8(testString)
	testAscii2.fromDecSixBit(testAscii.toDecSixBit())
	expected = b"".join([
	bytes([char]) if 0x20 <= char < 0x60 else b'?'
	for char in testString.upper()])
	assert testAscii2.toAscii8() == expected


def test_decSixBitLen():
	"""Dec Six Bit has a length of 6/8 byte per char
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert len(testAscii.toDecSixBit()) == 4 * 6


def test_ICL():
	"""ICL has edge cases and strange behaviours as shown by the
	inline for loop for the variable 'expected'
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii2 = AsciiBytes()
	testAscii.fromAscii8(testString)
	testAscii2.fromICL(testAscii.toICL())
	expected = b"".join([
	bytes([char]) if 0x20 <= char < 0x60 else b'?'
	for char in testString.upper()])
	assert testAscii2.toAscii8() == expected


def test_ICLLen():
	"""ICL has a length of 6/8 byte per char
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert len(testAscii.toICL()) == 4 * 6


def test_sixBitAscii():
	"""Six Bit Ascii has edge cases and strange behaviours as shown by the
	inline for loop for the variable 'expected'
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii2 = AsciiBytes()
	testAscii.fromAscii8(testString)
	testAscii2.fromSixBitAscii(testAscii.toSixBitAscii())
	expected = b"".join([
	bytes([char]) if 0x20 <= char < 0x60 else b'?'
	for char in testString.upper()])
	assert testAscii2.toAscii8() == expected


def test_sixBitAsciiLen():
	"""Six Bit Ascii has a length of 6/8 byte per char
	"""
	testString = generator(32)
	testAscii = AsciiBytes()
	testAscii.fromAscii8(testString)
	assert len(testAscii.toSixBitAscii()) == 4 * 6


miniencoding/tables.py
""" Convert charsets using a lookup table

https://docs.python.org/2.0/ref/strings.html
https://en.wikipedia.org/wiki/Six-bit_character_code#Types_of_six-bit_codes
https://en.wikipedia.org/wiki/GSM_03.38#GSM_7-bit_default_alphabet_and_extension_table_of_3GPP_TS_23.038_/_GSM_03.38
https://en.wikipedia.org/wiki/BCD_(character_encoding)
"""

from math import log2, ceil
from bitstring import BitArray

# CDC 1604: Magnetic tape BCD codes
CDC1604_MAGTAPE = "?1234567890#@??? /STUVWXYZ?,%???-JKLMNOPQR0$*???&ABCDEFGHI0.¤???"
# CDC 1604: Punched card codes
CDC1604_PUNCHCARD = "?1234567890=-??? /STUVWXYZ?,(???—JKLMNOPQR0$*???+ABCDEFGHI0.)???"
# CDC 1612: Printer codes (business applications)
CDC1612 = ":1234567890=≠≤![ /STUVWXYZ],(→≡~—JKLMNOPQR%$*↑↓>+ABCDEFGHI<.)≥?;"
# DEC SIXBIT
DEC_SIXBIT = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_"
# ECMA-1
EMCA1 = " \t\n\v\f\r\x0e\x0f()*+,-./0123456789:;<=>?\x00ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]\x1b\x7f"
# ICL Mainframes
ICL = "0123456789:;<=>? !\"#£%&'()*+,-./@ABCDEFGHIJKLMNOPQRSTUVWXYZ[$]↑←"
# SixBit ASCII (used by AIS)
SIXBIT = "@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_ !\"#$%&'()*+,-./0123456789:;<=>?"
# GOST 10859 § 6-bit code: with only Cyrillic upper case letters
GOST = "0123456789+-/,. ⏨↑()×=;[]*‘’≠<>:АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЫЬЭЮЯ\x7f"
# GSM 7-bit default alphabet and extension table of 3GPP TS 23.038 / GSM 03.38
GSM7 = ("@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ\x07ÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>"
+ "?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ§¿abcdefghijklmnopqrstuvwxyzäöñüà")
# ASCII
ASCII7 = bytes(range(0, 128)).decode("utf-8")
# IBM 48-character BCDIC code
IBM48 = " 1234567890#@????/STUVWXYZ?,%???-JKLMNOPQR?$*???&ABCDEFGHI?.⌑???"
# IBM 704 BCD code
IBM704 = "0123456789?#@???&ABCDEFGHI?.⌑???-JKLMNOPQR?$*??? /STUVWXYZ?,%???"
# IBM 7090/7094 character set
IBM7094 = IBM7090 = "0123456789?=\"???&ABCDEFGHI0.)???-JKLMNOPQR0$*??? /STUVWXYZ±,(???"
# IBM 1401 BCD code
IBM1401 = " 1234567890#@:>√¢/STUVWXYZ‡,%='\"-JKLMNOPQR!$*);Δ&ABCDEFGHI?.⌑(<⯒"
# GBCD code
GBCD = "0123456789[#@:>? ABCDEFGHI&.](<\\^JKLMNOPQR-$*);'+/STUVWXYZ_,%=\"!"
# Burroughs B5500 BCD code
BURROUGHS_B5500 = "0123456789#@?:>≥+ABCDEFGHI.[&(<←×JKLMNOPQR$*-);≤ /STUVWXYZ,%≠=]\""
# Code page 353
CP353 = " 1234567890#@:>√␢/STUVWXYZ‡,%γ\\⧻-JKLMNOPQR!#*];Δ&ABCDEFGHI?.⌑[<⯒"
# Code page 355
CP355 = " 1234567890#????@/STUVWXYZ‡,?γ??-JKLMNOPQR<$????&ABCDEFGHI).????"
# Code page 357
CP357 = " 1234567890=????'/STUVWXYZ‡,????-JKLMNOPQR!$????+ABCDEFGHI?.????"
# Code page 358
CP358 = " 1234567890'????!/STUVWXYZ‡,????-JKLMNOPQR<;????=ABCDEFGHI>.????"
# Code page 359/360
CP360 = CP359 = " 1234567890#????@/STUVWXYZ?,????-JKLMNOPQR?$????&ABCDEFGHI?.????"


def toUnicode(charset, bytestream):
	"""Convert text encoded with various charsets to unicode using a charset
	lookup table

	Args:
		charset (string): lookup table
		bytestream (bytes): sequence of bytes (split into bits of length
		log2(len(charset))) to convert

	Returns:
		string: unicode string
	"""
	unicode = []
	bitsize = ceil(log2(len(charset)))
	asciibits = BitArray(bytestream)
	for char in range(0, len(asciibits) // bitsize):
		unicode.append(charset[asciibits[char * bitsize:(char+1) * bitsize].uint])
	return "".join(unicode)


def toCharset(charset, string, toUpper=False):
	"""Convert unicode text to various charsets using a lookup table

	Args:
		charset (string): lookup table
		string (string): unicode string to convert
		toUpper (bool, optional): Make chars uppercase before converting
		(intended for 6 bit charsets or charsets that do not support uppercase).
		Defaults to False.

	Returns:
		bytes: sequence of bytes (split into bits of length
	"""
	asciibits = BitArray()
	cut = int(8 - log2(len(charset)))
	if toUpper:
		chars = chars.upper()
	for char in string:
		bits = charset.find(char)
		if bits < 0:
			if charset == GOST:
				bits = 25 # use '*' for GOST
			else:
				bits = charset.find("?")
		asciibits.append(BitArray(bytes([bits]))[cut:])
	return asciibits.tobytes()
